### ISSUE TITLE
Task-group identity and non-duplication constraints

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#1831]
The high-level requirement here is that we want to avoid the possibility that a user's task list includes the same task twice - which would be extremely confusing.
The definition of "same" task is partly TBD (see also #1829), since we should decide whether e.g. this includes (1) package extras [I guess yes], (2) python version [?], but that's not the major discussion point here.

cc @jluethi there is some complexity which we have not accounted for yet. In our previous discussions, we mostly thought about "private" and "public" tasks, but we also have to consider the intermediate case of "tasks shared with a given group".

Here are some examples where I'm not sure about the expected behaviors:

Preliminary steps:
1. `UserA` belongs to user groups `All`, `Team1` and `Collab3`. `UserB` belongs to user groups `All` and `Team2`
2. `UserA` collects task package X=1.2.3, and makes it available to `Team1` group
3. `UserB` collects task package X=1.2.3, and makes it available to `Team2` group
4. Up to here all is consistent, since either user only sees a single version of the task.

Starting from here, there are some operations that may break the high-level requirement above

5. If an admin wants to add `UserA` to `Team2`, this should fail because it would break the requirement above (i.e. `UserA` would have access to two versions of the same task). However
    * Checking for this is very cumbersome (I would need to check for all tasks that `UserA` has access to and all tasks that `Team2` has access to, and verify that they are disjoint sets).
    * It is counterintuitive that adding a user to a usergroup may fail due to some unrelated issue with their tasks.
6. If `UserA` wants to make their task package X=1.2.3 available for the `All` group, this should fail because it would break the requirement above (i.e. `UserB` would have access to two versions of the same task). However this means that we would need to check all task groups of all users of the target group (`All`, in this case), looking for overlaps with the current task group. This operation clearly scales badly with the number of tasks and users.
7. Similar to the previous point, if the user wants to change the group ownership of their taskgroup from `Team1` to `Collab3`, the same kind of checks need to be performed; it will be less expensive just becuase `Collab3` has less members than `All`, but it would scale in the same way

**Comment by jluethi** [comment]
> The high-level requirement here is that we want to avoid the possibility that a user's task list includes the same task twice - which would be extremely confusing.

We should avoid this as far as possible. If there are weird edge cases that cannot be prevented, we can let them through if needed (e.g. it's not a requirement that this can never ever happen, but it shouldn't be a common occurrence.

I'd phrase it less strongly:
For a given group (all, team1, collab3), every task package can only ever be collected once. And one cannot change the ownership of a task package to a group that already has the same version of that task package.

This could lead to all the complexity issues you have described with the groups. But the default should be that most tasks belong to all.

If we handle it somewhat cleverly on the front-end, maybe this conflict is actually fine:
If there are multiple versions of the same task package available to a user, only show the first according to the following order:
Task that only belong to the user
Task that belong to all
Task that belong to the groups in order of groups that were added (e.g. their groups is an ordered list, all is always first or treated separately)


> we should decide whether e.g. this includes (1) package extras [I guess yes], (2) python version [?]

(1) package extras: Yes, I don't think users will think about that
(2) python version: Same

=> I mostly think of it as "a given version of a given task package"

> cc @jluethi there is some complexity which we have not accounted for yet. In our previous discussions, we mostly thought about "private" and "public" tasks, but we also have to consider the intermediate case of "tasks shared with a given group".

Agreed. Also see above. I would accept a setup where we block this where obviously wrong (trying to collect a task package again for a group that already has access or to change the access to a group that already has that package). And if there are areas where we can raise useful warnings, we can do that.

> It is counterintuitive that adding a user to a usergroup may fail due to some unrelated issue with their tasks.

Agreed with this and the other examples. I'd rather we accept that in edge-cases, a task is available twice to a user.

**Comment by tcompa** [comment]
We re-assessed the situation with @mfranzon, based on the last message above.

## TLDR

Moving from a _global_ constraint (no user should see the "same" taskgroup twice) to two _local_ constraints (no usergroup should be associated to the same taskgroup twice, and no user should own the same taskgroup twice) is very helpful in reducing the complexities described above. The additional complexity (namely the disambiguation logic) would be enabled backend-side (ref #1850) but implemented client-side. Below we list what we plan to implement - and other valid options that for the moment we discarded.


## `TaskGroupV2` attributes

The first plan is to introduce the list of attributes listed below, for task groups. Note that many of those attributes are not specifically relevant for the current issue, but it's good to start to have them clear.

- `user_id: int`: ID of the user who created this task group (by POST-ing a single-task, or via any kind of task collection). This is the user who "owns" the task group, i.e. who can patch or delete it.
- `user_group_id: Optional[int] = None`: ID of the group of users who can access any task-group task (i.e. see it, insert it into a workflow, and run it).
    * If set, this user group should include the `user_id` user - upon creation.
    * TBD: Do we need to enforce this constraint forever? That is, is is allowed for an admin to remove the user from the usergroup even if this would mean that task-group owner does not belong to task-group usergroup?
- `origin: Literal["pypi", "wheel", "custom-collection", "single-task"]` - the specific names/options are TBD, but this is a useful information to track
- `pkg_name: str`
    * for pypi/wheel origins, this is a canonically-normalized package name
    * for custom task collection, this is either inferred or provided by the user
    * for custom single tasks, this is equal to task.name
- `version: Optional[str] = None`: optional, but always present for pypi/wheel origins
- `python_version: Optional[str] = None`: optional, but always present for pypi/wheel origins
- `venv_path: Optional[str] = None`: optional, but always present for pypi/wheel origins
- `pip_extras: Optional[str] = None`: only relevant for pypi/wheel
- `active: bool`, always True upon creation - can be edited later
- `timestamp_created` (required)

Additional fields may become useful for further lifecycle work, but are TBD for the moment:
- `timestamp_last_used` (updated upon every successful call to submit endpoint)
- `managed: bool`: similar to `origin`, this determines whether we can remove and recreate the venv.
- `pip_freeze: Optional[str]`: Information needed to recreate a venv.
- `lifecycle_activity_log: list[str]`: this could look like `["{timestamp1} TaskGroup created", "{timestamp2} TaskGroup deactivated", "{timestamp3} TaskGroup reactivated", ..]`

## Non-duplication constraints

Given a definition of "identical" task groups (see next section), we want to prevent the following situations from happening:
- Two "identical" task groups exist and are associated to the same user group.
- Two "identical" task groups exist and are associated to the same single user.

These two requirements are similar to each other, and they are both much simpler than the original global constraint, as can be seen in some examples based on those in https://github.com/fractal-analytics-platform/fractal-server/issues/1831#issue-2561255237:

- If `UserA` wants to create a _private_ task group `mypackage=1.2.3` but they already own one such task group, this operation is forbidden.
- If `UserA` wants to create a task group `mypackage=1.2.3` associated to group `Team1` but `Team1` is already associated to one such task group, this operation is forbidden.
- If an admin wants to modify some user/usergroup associations, this has no connection at all with the existence of tasks. This is much better than the counterintuitive situation that would result from a global constraint.
- If `UserA` wants to make their task group `mypackage=1.2.3` available for a given group (be it `All` or `Team1`) this still requires a preliminary check, but such check is very specific (as in: is there an identical task group associated to the given group? yes/no), and it does not involve looking at N users to see what their per-user view of tasks would be. Also here, we find this much more reasonable than the global constraint.
- If `UserA` wants to change the group ownership of one of their task groups from `Team1` to `All`, it's the same as in the previous point. A specific (local) check is required, but it does not involve a global view on the whole set of users.

Briefly: these new checks (one for the shared tasks and one for private ones) are much simpler and more intuitive, especially because they are not related to user-group memberships.

A mildly counterintuitive situation still exists, at the boundary between per-user and per-user-group constraints:
> Two identical task groups may exist and be visible to userA (one that userA created, and one to which userA has access because of some group ownership), but userA cannot be the creator of both. To rephrase it: userA cannot create two identical task groups, even if they would be associated to different user groups.
> This does not introduce any violation of the non-duplication constraint, but simply means that the two constraints are not fully independent and their overlap may also have an effect.


## `TaskGroupV2` identity

The first paln is that we define task-group "identity" via the combination of `pkg_name` and `version`. This is the basis for the non-duplication constraints above (no two task groups with the same _identity_ can be created for a given user of for a given user group).

Note that this is a broader definition than the `source`-based one we had in the past: it ignores the `origin` (pypi vs wheel vs other), the python version, the pip extras, the single-task identities. We list here some of the behaviors resulting from this choice of what "identity" means. If they do not match with expected behavior, we should adapt the identity definition.

1. Identity does not include `origin`: If I installed `fractal-tasks-core=1.2.3` via PyPI, I cannot re-install a custom version of it from a wheel file. Suggested solution: modify the package name in the local wheel, or make its version something like `1.2.3x`
2. Identity does not include `python_version`: I cannot install `fractal-tasks-core=1.2.3` both on Python 3.9 and 3.10. Suggested solution: none, generally speaking. If you really need both tasks, one of them should be renamed or installed only for a specific version.
3. Identity does not include `extras`: If I install `fractal-tasks-core=1.2.3` without the `fractal-tasks` extra, I cannot re-install it _with_ extras. Suggested solution: I have to delete the "wrong" task group first, and then re-install it with the appropriate extra.
4. Identity does not include single-task indentities: Say I installed a package `image-analysis` from a local wheel, where I had a `cellpose_segmentation` task. Then I update the wheel, adding a `sklearn_segmentation` task, and I try to install it again. This fails, because of the non-duplication constraint working on task _groups_ and not on the list of individual task. Suggested solution: I should bump the package version, or I should fully rename it if appropriate.
5. Identity does not include single-task indentities or `origin`: If I created a single task `image-analysis=1.2.3` (without task collection), then I cannot collect `image-analysis=1.2.3` from PyPI any more. Suggested solution: I should not name my single task as a PyPI package.
6. Same as the previous point holds when we replace a single-task creation with a custom task collection.
7. Identity does not include 
8. Any other idea of what could be counterintuitive with this definition of identity?

Note that the broad identity definition leads to a stricter non-duplication constraint. We may decide to make the identity definition more granular (e.g. by adding `origin`) and then accept more "duplications".

### Consequences for workflow export/import

Importing a workflow may now be based on different levels of strictness. In principle, all the following options could be supported (even though some do not really make sense):
- Upon import, look for a task group that has _all_ metadata identical to the exported one (including very specific ones like `timestamp_created`)
- Upon import, look for a task group that has some metadata identical to the exported one (package name, version, python version, pip extras).
- Upon import, look for a task group that has the "identity" metadata identical to the exported one (package name + version).
- Upon import, look for a task group that has part of the "identity" metadata identical to the exported one (e.g. identical package name, and version matching up to major.minor).
- Upon import, look for a task group that has only the package name matching with the exported one.


### A note about custom collections

Custom-collection task groups likely won't have all metadata that a standard-collection would create, e.g. they are not required to include a version. This means that it will be harder to re-collect them, because of the non-duplication constraints. This seems reasonable, given that the hybrid status of custom-collected tasks (somewhat midway between standard collections and custom tasks).

## A new requirement for task-collection endpoints

Even if the actual task collection runs in background (after the API call is over), the `TaskGroupV2` object should be created right away as part of the request - and not as part of the background operation. Reasons:
- If we postpone this object creation, two identical task collections could both start at the same time and lead to errors later.
- Creating the object right away also means that we can fail early (that is, respond with 422) e.g. in case of violations of the non-duplication constraints - rather than failing at the end of a long background operation.

This introduces a new requirement: we need to know the package version from within the endpoint, that is, before installing anything (same as in #1604).

- For wheel-based collections this is trivial, since the wheel already includes the version in its filename.
- For PyPI collections where the version is specified, this is also trivial.
- For PyPI collections not specifying the version (i.e. using "latest"), this would require an additional step like an API call (e.g. https://pypi.org/simple/fractal-tasks-core). I think we can make it such that it mostly works, and find out edge cases later.

Note that this would actually be a nice improvement in itself, independently on the needs described in this issue: we first find the target version, and then (in the background) try to install it.

### What do we mean with "edge cases"

Say that you ask for fractal-tasks-core "latest" for python 3.9, but we discontinued support for py3.9 starting from fractal-tasks-core 1.2.3. This means that the expected version you would get with `python3.9 -m pip install fractal-tasks-core` is 1.2.3. However, a naive check from the endpoint would set the version to whatever "latest" is today on PyPI (maybe 2.3.4, which only supports python 3.10). Thus the collection would start with 2.3.4, and fail because it's not available for Python 3.9. Another possible edge case may be coming from very different architecture for the server machine and cluster machines.

My first guess is that these edge cases will be rare and not very relevant, and if we hit them we may want to refine the pypi-information-retrieval step (e.g. by making use of the pypi response field `data-requires-python="&lt;4.0,&gt;=3.9"` which is already included).

## Discarded ideas: composite primary key

It'd be good to encode as much as possible of our constraints into the database structure, so that work on the API would only be needed to offer user-friendly errors - but not to enforce constraints.
In this case, this would correspond to making the set of `(pkg_name, version, user_group_id)` unique across TaskGroupV2 rows. This requires that `user_group_id` is non-nullable, or we would be being too strict (forbidding two different users from _privately_ installing the same package). But we'd need some additional change to include the case of "private" tasks (only visible to their creator).

### Option 1: Create single-user groups

- For every new user, we create a single-user group.
- We make these single-user groups special (through a flag), so that their membership is immutable. This introduces another (minor) discrepancy between API and database schemas.
- We represent the "private" task group as a task group which is associated to the single-user group corresponding to the task group owner.

This options does work but we are not fully ready to adopt it because it means that the number of user groups (and user/usergroup links) will keep growing with the number of users. In principle this is OK, but we suspect it could be cumbersome at some point.


### Option 2: Additional `private` flag

- If the task group is meant to be private, then we set `TaskGroupV2.private=True`.
- Task groups with `TaskGroupV2.private=False` do not appear as visible to anyone but their owner.
- Every task group is associated to a user group.
- If a task is private, the association to a user group is a "latent" information, which would only become useful if private is set to false.

Reasons to discard:
- If a task group has `private=True`, its association to the user group is meaningless at that precise point in time.
- There exists a fitting use case for this approach (provisionally setting a task group to private to perform an update, before sharing it with the same user group again), but it is not relevant in Fractal since our tasks are immutable (you are not meant to work on the code after it's been inserted in the db, unless it was a custom task collection).

**Comment by jluethi** [comment]
@tcompa Thanks for this additional pass on the design requirements. This looks super promising to me!

I'm going over all the points below. The only high-level point is that we may want to consider giving a user always access to view/run tasks that have their `user_id` set, not just via `user_group_id`. Besides that, some of the identity parts may require some additional thinking also for task life cycle, but all seems fine to me.

I'm a big fan of how clean this is & then more customizable imports we'd enable by this!

> On `user_group_id` If set, this user group should include the user_id user - upon creation.: TBD: Do we need to enforce this constraint forever? That is, is is allowed for an admin to remove the user from the usergroup even if this would mean that task-group owner does not belong to task-group usergroup?

Good question! Viewing & run access is purely handled via the `user_group_id`, while edit access is for admin or matchin `user_id`? It would be a bit funny to have tasks that a user is allowed to edit, but not to see...
See below why I think the owner should always be able to see & run their tasks.

> To rephrase it: userA cannot create two identical task groups, even if they would be associated to different user groups.
> This does not introduce any violation of the non-duplication constraint, but simply means that the two constraints are not fully independent and their overlap may also have an effect.

That is totally fine! We don't want users to collect the same task package multiple times and where we can reasonably stop this (e.g. the same user has already collected this, thus very likely has access to it*), we stop it.

*This does argue that a task "owner" (the user with the matching `user_id`) should always have access to use a task they created, even if they are not in the relevant group. Because they cannot create it again themselves. Which would be a weird loop:
1. UserA collects `mypackage=1.2.3` and associates it to `team2` (which they are not a member of)
2. UserA now does not see the task (=> cannot correct the team assignment), but also cannot collect the task package again.

> TaskGroupV2 identity
> The first plan is that we define task-group "identity" via the combination of pkg_name and version

That makes sense to me. 

- Identity does not include origin: Fine (& even good to avoid confusion!)
- Identity does not include python_version: Fine. Worst-case, can the task be deactivate and reinstalled with a different Python? Doesn't even need to be user exposed, we could make this an admin option in the lifecycle part.
- Identity does not include extras: Fine with delete & reinstall (or deactivate & admin action, see above)
- Identity does not include single-task indentities: Agreed, the version should get bumped once the package content changes. We support 0.1.0dev0, 0.1.0a0 etc., so that's very doable
- Identity does not include single-task identities or origin: Interesting edge-case. Also holds if 2 developers name their package the same for whls. Not much we can do about that I'd say. Out of scope for this discussion.
- Identity does not include *=> content missing?*

> Note that the broad identity definition leads to a stricter non-duplication constraint. We may decide to make the identity definition more granular (e.g. by adding origin) and then accept more "duplications".

Let's not start with that if we don't have to. Making the identity more granular later on sounds like it would remain possible, as the prior restrictions all hold.

------

> Consequences for workflow export/import

I like this!

> Upon import, look for a task group that has all metadata identical to the exported one (including very specific ones like timestamp_created)

This is super strict, but could become useful if Fractal is used in a more regulated environment.

> Upon import, look for a task group that has some metadata identical to the exported one (package name, version, python version, pip extras).
> Upon import, look for a task group that has the "identity" metadata identical to the exported one (package name + version).

That would be my default, maybe with warnings when Python version & pip extras are different, but still succeeding.

> Upon import, look for a task group that has part of the "identity" metadata identical to the exported one (e.g. identical package name, and version matching up to major.minor).
> Upon import, look for a task group that has only the package name matching with the exported one.

These could become interesting to enable better cross-portability. Scenarios: I created a workflow with `mypackage.task1=1.2.3` on `server1`. Now I want to use it on `server2`. Server 2 doesn't have `mypackage.task1=1.2.3`, but it has `mypackage.task1=1.2.5` or `mypackage.task1=1.3.0`. We can try the upgrade approach we have in Fractal web and see if it auto-updates.
I then try to import the workflow on `server3` which only has `mypackage.task1=1.2.1` => we should likely fail here and suggest that the task should be collected.


-----

> A note about custom collections
> Custom-collection task groups likely won't have all metadata that a standard-collection would create, e.g. they are not required to include a version. This means that it will be harder to re-collect them, because of the non-duplication constraints. This seems reasonable, given that the hybrid status of custom-collected tasks (somewhat midway between standard collections and custom tasks).

Agreed. Custom task collections really _SHOULD_ include a version. But in the end, it's a valve to support things the framework itself doesn't currently handle, so it doesn't need the same amount of support as pypi/whl collections.

----

> A new requirement for task-collection endpoints

This is good. Plus it means we can then show the task version in the interface during collection, also for whl collections. It was a confusion source in the past on why that wouldn't be shown :)


----

> Thus the collection would start with 2.3.4, and fail because it's not available for Python 3.9. 

I think that would be the right default. If you want a specific old version that works with Python 3.9, specify the version. If you wanted latest and it doesn't work with your Python, you should get an error.

**Comment by tcompa** [comment]
Two high-level comments, without going into details (for now):
1. Yes, a task is _always_ visible to the user who creates it. Access-control rules defined within https://github.com/fractal-analytics-platform/fractal-server/issues/1803 are already implemented in `main`.
2. We did not consider the value of `active` to be relevant for the non-duplication constraint, but I guess we can rephrase the constraint as "there cannot be two _active_ task groups such that ..". The downside of this restriction has to do with the following scenario:
    * someone collects fractal-tasks-core 0.1.2
    * someone else uses those tasks in their workfloes
    * owner or admin sets fractal-tasks-core 0.1.2 to `active=False`, in view of deprecation or in order to push users to more recent versions
    * another user is now allowed to collect fractal-tasks-core 0.1.2
    * if this happens, admin cannot re-activate their original package any more (as it could break a constraint)
    * users ask to reactivate the package which is needed to re-run their previously-working jobs

**Comment by jluethi** [comment]
On tasks being visible to creator: Great, then all those concerns are covered!

Hmm, not sure whether I would suggest using active that way. In:

> Worst-case, can the task be deactivate and reinstalled with a different Python?

I mostly meant: Task is deactivated, environment is removed. Admin changes Python version of the task package in the db. Then recollects the task
=> Very edge-case workaround for when there is no new version of a task package and it needs to be reinstalled with a different Python. Not something I'd expose to users


For the actual scenario above:
another user is now allowed to collect fractal-tasks-core 0.1.2 => that's not the goal of deactivation, right? We can decide who can reactivate when, but "re-collecting 0.1.2" to me should mean reactivating it, if there is a deactivated version.
(at least for tasks available to the ALL group, we can discuss if this gets more complex in the other scenarios)

**Comment by tcompa** [comment]
> On tasks being visible to creator: Great, then all those concerns are covered!
> 
> Hmm, not sure whether I would suggest using active that way. In:
> 
> > Worst-case, can the task be deactivate and reinstalled with a different Python?
> 
> I mostly meant: Task is deactivated, environment is removed. Admin changes Python version of the task package in the db. Then recollects the task => Very edge-case workaround for when there is no new version of a task package and it needs to be reinstalled with a different Python. Not something I'd expose to users

This is all fine, then. 
(I first had read your suggestion as "let's have a deactivate&recollect feature", but this would not have worked because `active` is not part of the task-group identity).


> For the actual scenario above: another user is now allowed to collect fractal-tasks-core 0.1.2 => that's not the goal of deactivation, right?

Agreed, this is not the goal of deactivation. To put it more clearly: activation/deactivation operation do not have to do with either task-group ownership or non-duplication constraints.

> We can decide who can reactivate when, but "re-collecting 0.1.2" to me should mean reactivating it, if there is a deactivated version. (at least for tasks available to the ALL group, we can discuss if this gets more complex in the other scenarios)

This ("re-collecting somehow corresponds to re-activating) sounds reasonable. I'd say we start with an informative error message, and we ca re-discuss it once we get a bit further in the lifecycle discussion.

**Comment by tcompa** [comment]
In my understanding, all points discussed above are reasonably defined - I'll start to unpack them into separate issues so that we can proceed with implementation.
